### PR TITLE
Dropout kernel and function body expansion updates.

### PR DESCRIPTION
**Pull request by codemzs**
The change addresses two issues:

- **Function body**
  - This change enables the handling of function bodies that require function proto context in ORT by creating function body at Node Verification stage.
  - Fixes bugs around function body expansion that were causing segmentation faults and index out of bound exceptions.
  - Adds tests for NegativeLikelihoodLoss forward kernel (defined by function). 

- **Dropout-12 kernel**
  - Pursuant to https://github.com/onnx/onnx/pull/2725 this PR introduces `training_mode` input to Dropout-12 kernel to repurpose this kernel for both inferencing and training without breaking the existing inferencing behavior where default probability is zero. The motivation for the change was to change the spec's default probability to 0.5 to make it inline with PyTorch, TensorFlow and ONNX Runtime Training.
  - Updates ORT-ONNX test infrastructure to handle dropout tests by counting the number of zeros in the output as opposed to doing absolute comparisions which will _almost_ never match because mask generated by ORT will be different from that of ONNX model baseline (generated by python based reference implementation that use np.random distribution) 
  - Separates out TrainableDropout because existing models using it will pass `training_mode` input and instead will expect it to be used for training directly. 

Please note CI builds are failing due to below tests not passing but I believe these failures are **not** related to my change, please review with confidence:

- ParallelExecutor.TestStatusPropagation
- ParallelExecutorThreadPoolTests/ParallelExecutorThreadPoolTest.TestNullInterOpThreadPool/0, where GetParam() = 1
- ParallelExecutorThreadPoolTests/ParallelExecutorThreadPoolTest.TestNullInterOpThreadPool/1, where GetParam() = 0

### CODE REVIEW

**Comment by SherlockNoMad** [comment]
Thanks for the prompt fix. 
Please include "Training" label to the PR, if it requires review from training team.

File: onnxruntime/core/graph/function.cc [hunk 1]
```diff
@@ -226,7 +226,7 @@ static std::unordered_map<std::string, int> GetOpsetVersionMap(const ONNX_NAMESP
 
 FunctionImpl::FunctionImpl(const onnxruntime::Graph& graph,
                            const onnxruntime::NodeIndex& node_index,
-                           const ONNX_NAMESPACE::FunctionProto& onnx_func_proto,
```
Comment (SherlockNoMad) [inline onnxruntime/core/graph/function.cc:229]: pass by const ref ?

File: onnxruntime/core/graph/function.cc [hunk 2]
```diff
@@ -268,12 +268,12 @@ FunctionImpl::FunctionImpl(const onnxruntime::Graph& graph,
     }
     int i = 0;
     for (auto& input : cached_op_schema->inputs()) {
-      op_schema_->Input(i, input.GetName(), input.GetDescription(), input.GetTypeStr());
+      op_schema_->Input(i, input.GetName(), input.GetDescription(), input.GetTypeStr(), input.GetOption());
```
Comment (SherlockNoMad) [inline onnxruntime/core/graph/function.cc:271]: How about is_homogeneous and min_arity? Any harm to add them ?

File: onnxruntime/core/graph/graph.cc [hunk 3]
```diff
@@ -1964,8 +1965,17 @@ Status Graph::VerifyNodeAndOpMatch(const ResolveOptions& options) {
         node.op_ = nullptr;
       }
 
-      if (node.op_ && node.op_->HasFunction()) {
-        auto onnx_function_proto = node.op_->GetFunction();
+      if (node.op_ && (node.op_->HasFunction() || node.op_->HasContextDependentFunction())) {
+        const onnx::FunctionProto* onnx_function_proto;
+        onnx::FunctionProto temp;
```
Comment (SherlockNoMad) [inline onnxruntime/core/graph/graph.cc:1970]: It seems to me that we don't need the temp here?
onnx::FunctionProto onnx_function_proto;

line 1973 
node.op_->BuildContextDependentFunction(function_body_ctx, onnx_function_proto);

File: onnxruntime/test/providers/provider_test_utils.cc [hunk 4]
```diff
@@ -798,6 +798,26 @@ void OpTester::Run(
               }
             }
 
+            // Review(codemzs): Recurses only one level.
+            if (!kci) {
```
Comment (SherlockNoMad) [inline onnxruntime/test/providers/provider_test_utils.cc:802]: seems duplicated with line 791 ?

**Comment by SherlockNoMad** [comment]
+ Rama to review

File: onnxruntime/core/graph/function.cc [hunk 5]
```diff
@@ -300,8 +300,8 @@ FunctionImpl::FunctionImpl(const onnxruntime::Graph& graph,
   //TODO: set correct domain and version
   domain_to_version[onnxruntime::kOnnxDomain] = static_cast<int>(onnx_func_proto_.since_version());
   auto& function_body_graph = body_.MainGraph();
-  std::vector<const NodeArg*> graph_inputs(onnx_func_proto_.input_size(), nullptr),
-      graph_outputs(onnx_func_proto_.output_size(), nullptr);
+  std::vector<const NodeArg*> graph_inputs(node_in_parent_graph->InputDefs().size() /*onnx_func_proto_.input_size()*/, nullptr),
```
[the remark this answers is not in the document]
Comment (codemzs) [reply onnxruntime/core/graph/function.cc:303]: Yep, forgot about it, this commented stuff is actually what was causing out of bounds exception.

File: onnxruntime/core/graph/graph.cc (hunk 3 above)
Comment (codemzs) [reply onnxruntime/core/graph/graph.cc:1970]: yea we don't it was for debugging when I was investigating crash dumps ... naive approach.

File: onnxruntime/core/graph/function.cc (hunk 2 above)
Comment (codemzs) [reply onnxruntime/core/graph/function.cc:271]: sure, I add options because it was actually a bug and causing segmentation fault. 

---
In reply to: [411612229](https://github.com/microsoft/onnxruntime/pull/3567#discussion_r411612229) [](ancestors = 411612229)

File: orttraining/orttraining/test/training_ops/cuda/cross_entropy_test.cc [hunk 6]
```diff
@@ -290,7 +290,7 @@ static void TestSoftmaxCrossEntropyLoss(const std::vector<int64_t>* X_dims,
   std::vector<int64_t> index_data = random.Uniform<int64_t>(*index_dims, 0.0f, static_cast<float>((*X_dims)[1]));
   //Add one data point that has ignore_index.
   if (index_data.size() > 0) {
-    index_data[0] = ignore_index;
+    index_data[0] = static_cast<int32_t>(ignore_index);
```
[the remark this answers is not in the document]
Comment (codemzs) [reply orttraining/orttraining/test/training_ops/cuda/cross_entropy_test.cc:293]: It doesn't make sense to cast to int32 and I'm not sure why I did that. The vector is of type in64 and so is ignore_index can;'t remember why I did this. Anyways reverted.

---
In reply to: [411617087](https://github.com/microsoft/onnxruntime/pull/3567#discussion_r411617087) [](ancestors = 411617087)

File: onnxruntime/test/providers/provider_test_utils.cc (hunk 4 above)
Comment (codemzs) [reply onnxruntime/test/providers/provider_test_utils.cc:807]: I added this while trying to get grad test working only to realize the backward expansion setup is not in place hence there is no point adding this change., I will revert.,

---
In reply to: [411621655](https://github.com/microsoft/onnxruntime/pull/3567#discussion_r411621655) [](ancestors = 411621655)

File: onnxruntime/test/util/compare_ortvalue.cc [hunk 7]
```diff
@@ -168,15 +223,16 @@ std::pair<COMPARE_RESULT, std::string> CompareBFloat16Result(const Tensor& outva
 
 std::pair<COMPARE_RESULT, std::string> CompareTwoTensors(const Tensor& outvalue, const Tensor& expected_tensor,
                                                          double per_sample_tolerance,
-                                                         double relative_per_sample_tolerance, bool post_processing) {
+                                                         double relative_per_sample_tolerance, bool post_processing,
+                                                         bool dropout_test = false) {
```
Comment (SherlockNoMad) [inline onnxruntime/test/util/compare_ortvalue.cc:227]: This is a general purposed utility function. Introducing a flag just to make it support a single special case, Dropout, is not the best idea...

**Comment by codemzs** [comment]
@SherlockNoMad I thought you said to add “core” label when you touch graph.cc?

File: onnxruntime/test/util/compare_ortvalue.cc [hunk 8]
```diff
@@ -106,17 +135,43 @@ std::pair<COMPARE_RESULT, std::string> CompareFloatResult(const Tensor& outvalue
 }
 
 template <typename T>
-std::pair<COMPARE_RESULT, std::string> IsResultExactlyMatch(const Tensor& outvalue, const Tensor& expected_value) {
+std::pair<COMPARE_RESULT, std::string> IsResultExactlyMatch(const Tensor& outvalue, const Tensor& expected_value,
+    bool dropout = false) {
```
Comment (SherlockNoMad) [inline onnxruntime/test/util/compare_ortvalue.cc:139]: Same here. 
You can introduce another utility function to handle the special case if need.

File: onnxruntime/test/onnx/TestCase.h [hunk 9]
```diff
@@ -19,6 +19,7 @@ class ValueInfoProto;
 //One test case can contain multiple test data(input/output pairs)
 class ITestCase {
  public:
+  ITestCase() : dropout_test(false){};
```
Comment (SherlockNoMad) [inline onnxruntime/test/onnx/TestCase.h:22]: This is a test case interface... ITestCase... 
Introducing such flag can't be a good design pattern

File: onnxruntime/test/util/compare_ortvalue.cc (hunk 7 above)
Comment (codemzs) [reply onnxruntime/test/util/compare_ortvalue.cc:227]: Are you ok it I passed an optional function pointer to do the comparison and set result in this function? It will be a little cleaner and if no function pointer is passed then it can apple default element by element comparison

File: orttraining/orttraining/test/training_ops/cpu/nn/dropout_op_test.cc [hunk 10]
```diff
@@ -34,7 +34,7 @@ const Tensor& FetchTensor(const OrtValue& ort_value) {
 }
 
 void RunDropoutTest(const char* op, const bool use_mask, const std::vector<int64_t>& input_shape, float ratio = -1,
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/training_ops/cpu/nn/dropout_op_test.cc:36]: See line 76. 
it's already using counting 0s and compare it with the ratio.

Comment (codemzs) [reply orttraining/orttraining/test/training_ops/cpu/nn/dropout_op_test.cc:36]: I’m aware but what is your point? These are ORT tests not ONNX tests.

Comment (SherlockNoMad) [reply orttraining/orttraining/test/training_ops/cpu/nn/dropout_op_test.cc:36]: You can enrich output_verifier if needed. 
But modifying optest and comparer utility is too intrusive.

File: orttraining/orttraining/training_ops/cpu/nn/dropout_op.cc [hunk 11]
```diff
@@ -61,6 +61,7 @@ Status Dropout<T1, T2>::Compute(OpKernelContext* context) const {
 
   const Tensor* ratio = context->Input<Tensor>(1);  // optional
   const float ratio_value = GetRatioOrDefault<T2>(ratio);
+  ORT_ENFORCE(ratio_value >= 0.0f && ratio_value < 1.0f);
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/nn/dropout_op.cc:64]: I believe this is already checked within GetRatioOrDefault.

File: orttraining/orttraining/training_ops/cpu/nn/dropout_op.h [hunk 12]
```diff
@@ -10,27 +10,54 @@ namespace onnxruntime {
...
+    trainable_dropout_ = false;
   }
 
   Status Compute(OpKernelContext* context) const override;
 
  private:
   mutable std::unique_ptr<RandomGenerator> generator_;
+
+ protected: 
+  bool trainable_dropout_;
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/nn/dropout_op.h:30]: This can be a class template parameter... See softmax/logsoftmax for ref

This way, do you don't need to redefine another TrainableDropout Class, 
Dropout class can still have "final" property.

Comment (codemzs) [reply orttraining/orttraining/training_ops/cpu/nn/dropout_op.h:30]: This is a great idea! Thank you!

File: orttraining/orttraining/training_ops/cuda/nn/dropout.cc [hunk 13]
```diff
@@ -47,11 +48,30 @@ Status Dropout<T1, T2>::ComputeInternal(OpKernelContext* context) const {
 
   //Get Y_data
   auto Y = context->Output(0, shape);
+  ORT_ENFORCE(Y->Shape() == shape, "X and Y should have the same shape");
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/nn/dropout.cc:51]: The check seems redundant... 
Y is created according to shape one line above ... 

Usually, we check the shape for inputs if there is any requirement

**Comment by edgchen1** [comment]
deleting ort_training, please rebase work on master
